### ISSUE TITLE
"Error reducing result rows" exception when downloading full results

### ISSUE THREAD

**Issue by vreon** [metabase/metabase#12339]
**Describe the bug**

Using the Metabase UI to "download full results" on a question produces an incomplete file whose last line is a JSON-formatted stack trace.

**Logs**

Here's the error that gets printed to the file (indented with jq). I also redacted the query itself, but let me know if it could be relevant.

```json
{
  "database_id": 2,
  "started_at": "2020-04-16T19:42:37.099433Z",
  "via": [
    {
      "status": "failed",
      "class": "class clojure.lang.ExceptionInfo",
      "error": "Error reducing result rows",
      "stacktrace": [
        "--> query_processor.context.default$default_reducef$fn__37885.invoke(default.clj:61)",
        "query_processor.context.default$default_reducef.invokeStatic(default.clj:58)",
        "query_processor.context.default$default_reducef.invoke(default.clj:49)",
        "query_processor.context$reducef.invokeStatic(context.clj:69)",
        "query_processor.context$reducef.invoke(context.clj:62)",
        "query_processor.context.default$default_runf$respond_STAR___37889.invoke(default.clj:70)",
        "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:392)",
        "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:377)",
        "driver.sql_jdbc$fn__72719.invokeStatic(sql_jdbc.clj:50)",
        "driver.sql_jdbc$fn__72719.invoke(sql_jdbc.clj:48)",
        "query_processor.context$executef.invokeStatic(context.clj:59)",
        "query_processor.context$executef.invoke(context.clj:48)",
        "query_processor.context.default$default_runf.invokeStatic(default.clj:69)",
        "query_processor.context.default$default_runf.invoke(default.clj:67)",
        "query_processor.context$runf.invokeStatic(context.clj:45)",
        "query_processor.context$runf.invoke(context.clj:39)",
        "query_processor.reducible$pivot.invokeStatic(reducible.clj:34)",
        "query_processor.reducible$pivot.invoke(reducible.clj:31)",
        "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__43022.invoke(mbql_to_native.clj:26)",
        "query_processor.middleware.check_features$check_features$fn__42336.invoke(check_features.clj:42)",
        "query_processor.middleware.optimize_datetime_filters$optimize_datetime_filters$fn__43187.invoke(optimize_datetime_filters.clj:133)",
        "query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__47085.invoke(wrap_value_literals.clj:137)",
        "query_processor.middleware.annotate$add_column_info$fn__40965.invoke(annotate.clj:577)",
        "query_processor.middleware.permissions$check_query_permissions$fn__42211.invoke(permissions.clj:64)",
        "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__43686.invoke(pre_alias_aggregations.clj:40)",
        "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__42409.invoke(cumulative_aggregations.clj:61)",
        "query_processor.middleware.resolve_joins$resolve_joins$fn__44218.invoke(resolve_joins.clj:183)",
        "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__39152.invoke(add_implicit_joins.clj:245)",
        "query_processor.middleware.limit$limit$fn__43008.invoke(limit.clj:38)",
        "query_processor.middleware.format_rows$format_rows$fn__42989.invoke(format_rows.clj:81)",
        "query_processor.middleware.desugar$desugar$fn__42475.invoke(desugar.clj:22)",
        "query_processor.middleware.binning$update_binning_strategy$fn__41509.invoke(binning.clj:229)",
        "query_processor.middleware.resolve_fields$resolve_fields$fn__42017.invoke(resolve_fields.clj:24)",
        "query_processor.middleware.add_dimension_projections$add_remapping$fn__38688.invoke(add_dimension_projections.clj:270)",
        "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__38908.invoke(add_implicit_clauses.clj:147)",
        "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__39301.invoke(add_source_metadata.clj:105)",
        "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__43883.invoke(reconcile_breakout_and_order_by_bucketing.clj:98)",
        "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__41150.invoke(auto_bucket_datetimes.clj:125)",
        "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__42064.invoke(resolve_source_table.clj:46)",
        "query_processor.middleware.parameters$substitute_parameters$fn__43668.invoke(parameters.clj:97)",
        "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__42116.invoke(resolve_referenced.clj:80)",
        "query_processor.middleware.expand_macros$expand_macros$fn__42731.invoke(expand_macros.clj:158)",
        "query_processor.middleware.add_timezone_info$add_timezone_info$fn__39332.invoke(add_timezone_info.clj:15)",
        "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__46969.invoke(splice_params_in_response.clj:32)",
        "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__43894$fn__43898.invoke(resolve_database_and_driver.clj:33)",
        "driver$do_with_driver.invokeStatic(driver.clj:61)",
        "driver$do_with_driver.invoke(driver.clj:57)",
        "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__43894.invoke(resolve_database_and_driver.clj:27)",
        "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__42937.invoke(fetch_source_query.clj:243)",
        "query_processor.middleware.store$initialize_store$fn__46978$fn__46979.invoke(store.clj:11)",
        "query_processor.store$do_with_store.invokeStatic(store.clj:46)",
        "query_processor.store$do_with_store.invoke(store.clj:40)",
        "query_processor.middleware.store$initialize_store$fn__46978.invoke(store.clj:10)",
        "query_processor.middleware.cache$run_query_with_cache.invokeStatic(cache.clj:181)",
        "query_processor.middleware.cache$run_query_with_cache.invoke(cache.clj:172)",
        "query_processor.middleware.cache$maybe_return_cached_results$fn__41993.invoke(cache.clj:207)",
        "query_processor.middleware.validate$validate_query$fn__46987.invoke(validate.clj:10)",
        "query_processor.middleware.normalize_query$normalize$fn__43035.invoke(normalize_query.clj:22)",
        "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__39170.invoke(add_rows_truncated.clj:36)",
        "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__46954.invoke(results_metadata.clj:124)",
        "query_processor.middleware.constraints$add_default_userland_constraints$fn__42352.invoke(constraints.clj:42)",
        "query_processor.middleware.process_userland_query$process_userland_query$fn__43757.invoke(process_userland_query.clj:136)",
        "query_processor.middleware.catch_exceptions$catch_exceptions$fn__42295.invoke(catch_exceptions.clj:166)",
        "query_processor.reducible$async_qp$qp_STAR___37971$thunk__37972.invoke(reducible.clj:100)",
        "query_processor.reducible$async_qp$qp_STAR___37971$fn__37974.invoke(reducible.clj:105)"
      ],
      "error_type": "qp",
      "ex-data": {
        "type": "qp"
      }
    }
  ],
  "error_type": "qp",
  "json_query": {
    "constraints": null,
    "type": "native",
    "middleware": {
      "skip-results-metadata?": true
    },
    "native": {
      "query": "[redacted]"
    },
    "database": 2,
    "parameters": null,
    "async?": true,
    "cache-ttl": 3975
  },
  "status": "failed",
  "class": "class java.lang.AssertionError",
  "stacktrace": [
    "clojure.core.async.impl.channels.ManyToManyChannel.put_BANG_(channels.clj:152)",
    "clojure.core.async$put_BANG_.invokeStatic(async.clj:165)",
    "clojure.core.async$put_BANG_.invoke(async.clj:158)",
    "--> query_processor.middleware.cache$save_results_xform$fn__41953.invoke(cache.clj:110)",
    "query_processor.middleware.format_rows$format_rows_xform$fn__42982.invoke(format_rows.clj:75)",
    "query_processor.middleware.limit$limit_xform$fn__43005.invoke(limit.clj:25)",
    "query_processor.reducible$combine_additional_reducing_fns$fn__37994.invoke(reducible.clj:211)",
    "query_processor.reducible$reducible_rows$reify__37987.reduce(reducible.clj:156)",
    "query_processor.context.default$default_reducef$fn__37885.invoke(default.clj:59)",
    "query_processor.context.default$default_reducef.invokeStatic(default.clj:58)",
    "query_processor.context.default$default_reducef.invoke(default.clj:49)",
    "query_processor.context$reducef.invokeStatic(context.clj:69)",
    "query_processor.context$reducef.invoke(context.clj:62)",
    "query_processor.context.default$default_runf$respond_STAR___37889.invoke(default.clj:70)",
    "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:392)",
    "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:377)",
    "driver.sql_jdbc$fn__72719.invokeStatic(sql_jdbc.clj:50)",
    "driver.sql_jdbc$fn__72719.invoke(sql_jdbc.clj:48)",
    "query_processor.context$executef.invokeStatic(context.clj:59)",
    "query_processor.context$executef.invoke(context.clj:48)",
    "query_processor.context.default$default_runf.invokeStatic(default.clj:69)",
    "query_processor.context.default$default_runf.invoke(default.clj:67)",
    "query_processor.context$runf.invokeStatic(context.clj:45)",
    "query_processor.context$runf.invoke(context.clj:39)",
    "query_processor.reducible$pivot.invokeStatic(reducible.clj:34)",
    "query_processor.reducible$pivot.invoke(reducible.clj:31)",
    "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__43022.invoke(mbql_to_native.clj:26)",
    "query_processor.middleware.check_features$check_features$fn__42336.invoke(check_features.clj:42)",
    "query_processor.middleware.optimize_datetime_filters$optimize_datetime_filters$fn__43187.invoke(optimize_datetime_filters.clj:133)",
    "query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__47085.invoke(wrap_value_literals.clj:137)",
    "query_processor.middleware.annotate$add_column_info$fn__40965.invoke(annotate.clj:577)",
    "query_processor.middleware.permissions$check_query_permissions$fn__42211.invoke(permissions.clj:64)",
    "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__43686.invoke(pre_alias_aggregations.clj:40)",
    "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__42409.invoke(cumulative_aggregations.clj:61)",
    "query_processor.middleware.resolve_joins$resolve_joins$fn__44218.invoke(resolve_joins.clj:183)",
    "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__39152.invoke(add_implicit_joins.clj:245)",
    "query_processor.middleware.limit$limit$fn__43008.invoke(limit.clj:38)",
    "query_processor.middleware.format_rows$format_rows$fn__42989.invoke(format_rows.clj:81)",
    "query_processor.middleware.desugar$desugar$fn__42475.invoke(desugar.clj:22)",
    "query_processor.middleware.binning$update_binning_strategy$fn__41509.invoke(binning.clj:229)",
    "query_processor.middleware.resolve_fields$resolve_fields$fn__42017.invoke(resolve_fields.clj:24)",
    "query_processor.middleware.add_dimension_projections$add_remapping$fn__38688.invoke(add_dimension_projections.clj:270)",
    "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__38908.invoke(add_implicit_clauses.clj:147)",
    "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__39301.invoke(add_source_metadata.clj:105)",
    "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__43883.invoke(reconcile_breakout_and_order_by_bucketing.clj:98)",
    "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__41150.invoke(auto_bucket_datetimes.clj:125)",
    "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__42064.invoke(resolve_source_table.clj:46)",
    "query_processor.middleware.parameters$substitute_parameters$fn__43668.invoke(parameters.clj:97)",
    "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__42116.invoke(resolve_referenced.clj:80)",
    "query_processor.middleware.expand_macros$expand_macros$fn__42731.invoke(expand_macros.clj:158)",
    "query_processor.middleware.add_timezone_info$add_timezone_info$fn__39332.invoke(add_timezone_info.clj:15)",
    "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__46969.invoke(splice_params_in_response.clj:32)",
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__43894$fn__43898.invoke(resolve_database_and_driver.clj:33)",
    "driver$do_with_driver.invokeStatic(driver.clj:61)",
    "driver$do_with_driver.invoke(driver.clj:57)",
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__43894.invoke(resolve_database_and_driver.clj:27)",
    "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__42937.invoke(fetch_source_query.clj:243)",
    "query_processor.middleware.store$initialize_store$fn__46978$fn__46979.invoke(store.clj:11)",
    "query_processor.store$do_with_store.invokeStatic(store.clj:46)",
    "query_processor.store$do_with_store.invoke(store.clj:40)",
    "query_processor.middleware.store$initialize_store$fn__46978.invoke(store.clj:10)",
    "query_processor.middleware.cache$run_query_with_cache.invokeStatic(cache.clj:181)",
    "query_processor.middleware.cache$run_query_with_cache.invoke(cache.clj:172)",
    "query_processor.middleware.cache$maybe_return_cached_results$fn__41993.invoke(cache.clj:207)",
    "query_processor.middleware.validate$validate_query$fn__46987.invoke(validate.clj:10)",
    "query_processor.middleware.normalize_query$normalize$fn__43035.invoke(normalize_query.clj:22)",
    "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__39170.invoke(add_rows_truncated.clj:36)",
    "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__46954.invoke(results_metadata.clj:124)",
    "query_processor.middleware.constraints$add_default_userland_constraints$fn__42352.invoke(constraints.clj:42)",
    "query_processor.middleware.process_userland_query$process_userland_query$fn__43757.invoke(process_userland_query.clj:136)",
    "query_processor.middleware.catch_exceptions$catch_exceptions$fn__42295.invoke(catch_exceptions.clj:166)",
    "query_processor.reducible$async_qp$qp_STAR___37971$thunk__37972.invoke(reducible.clj:100)",
    "query_processor.reducible$async_qp$qp_STAR___37971$fn__37974.invoke(reducible.clj:105)"
  ],
  "context": "csv-download",
  "error": "Assert failed: No more than 1024 pending puts are allowed on a single channel. Consider using a windowed buffer.\n(< (.size puts) impl/MAX-QUEUE-SIZE)",
  "row_count": 0,
  "running_time": 0,
  "data": {
    "rows": [],
    "cols": []
  }
}
```

**To Reproduce**

1. Find a question that returns a large result set (in our case, ~50k rows)
2. Click on "Download full results" and select CSV
3. Open the resulting CSV
4. The CSV will be incomplete (in our case, only a few thousand rows, exact number varies on each attempt) and the final line is a stack trace

**Expected behavior**

Expected the full results to be present in the downloaded file.

**Metabase Diagnostic Info**

```json
{
  "browser-info": {
    "language": "en-US",
    "platform": "Linux x86_64",
    "userAgent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.7 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.5+10",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.5",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.5+10",
    "os.name": "Linux",
    "os.version": "3.10.0-1062.9.1.el7.x86_64",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "mysql",
      "googleanalytics",
      "mongo"
    ],
    "hosting-env": "unknown",
    "application-database": "mysql",
    "application-database-details": {
      "database": {
        "name": "MySQL",
        "version": "5.7.23-25"
      },
      "jdbc-driver": {
        "name": "MariaDB Connector/J",
        "version": "2.5.1"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2020-04-10",
      "tag": "v0.35.2",
      "branch": "release-0.35.x",
      "hash": "f3b2857"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

**Severity**

Somewhat annoying, but not a showstopper. Our Metabase users are currently asking developers to run such reports using the MySQL CLI in the interim.

**Additional context**

Recent changes to our Metabase instance:

- Upgraded from 0.31.2 to 0.35.2
- Enabled caching (min duration: 30s, TTL multiplier: 100)

Unfortunately, I'm not sure if this error was also present before either of those changes.

**Comment by flamber** [comment]
Duplicate of #12251 - is fixed and will be part of 0.35.3

**Comment by sbelak** [comment]
I don't think this is a dupe of 12251. This seems to be caused by some interaction of transducers and async.

**Comment by sbelak** [comment]
Fixed in #12407
